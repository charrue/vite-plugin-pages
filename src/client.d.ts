declare module 'vite-plugin-pages/client' {
  import { RouteRecordRaw } from 'vue-router'
  declare const routes: RouteRecordRaw[]
  export default routes
}
