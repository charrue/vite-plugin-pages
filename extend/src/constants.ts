export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
