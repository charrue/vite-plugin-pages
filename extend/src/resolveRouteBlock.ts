export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
export const MACROS_FN_STR = `function ${MACROS_KEY}(config) { return config };`

export const isJsx = (filePath: string) => {
  return filePath.endsWith('.jsx') || filePath.endsWith('.tsx')
}

export const resolveRouteBlock = (content: string, filePath: string) => {
  if (isJsx(filePath) && content.includes(MACROS_KEY)) {
    const body = content.match(MACROS_KEY_RE)?.[0]
    if (body) {
      // eslint-disable-next-line no-new-func
      const routeConfig = new Function(`${MACROS_FN_STR} return ${body}`)()
      return routeConfig
    }
  }
}
