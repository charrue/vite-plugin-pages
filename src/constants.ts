export const MODULE_IDS = ['pages-generated', 'virtual:generated-pages']
export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages'
