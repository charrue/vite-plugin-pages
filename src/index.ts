import { resolve } from 'path'
import type { Plugin, ResolvedConfig, ModuleNode } from 'vite'
import { Route, ResolvedOptions, UserOptions, PageDirOptions } from './types'
import { getFilesFromPath } from './files'
import { generateRoutes, generateClientCode, updateRouteFromHMR } from './generate'
import { debug, normalizePath } from './utils'
import { parseVueRequest } from './query'

const ID = 'pages-generated'

function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  const {
    pagesDir = ['src/pages'],
    extensions = ['vue', 'js'],
    importMode = 'async',
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
  } = userOptions

  const root = process.cwd()
  const pagesDirPaths: string[] = []

  let pagesDirOptions: PageDirOptions[] = []

  if (typeof pagesDir === 'string') {
    pagesDirOptions = pagesDirOptions.concat({ dir: pagesDir, baseRoute: '' })
  }
  else {
    for (const dir of pagesDir) {
      if (typeof dir === 'string')
        pagesDirOptions = pagesDirOptions.concat({ dir, baseRoute: '' })
      else if (dir as PageDirOptions)
        pagesDirOptions = pagesDirOptions.concat(dir)
    }
  }

  return Object.assign(
    {},
    {
      routeBlockLang,
      root,
      pagesDir,
      pagesDirPaths,
      pagesDirOptions,
      extensions,
      importMode,
      exclude,
      syncIndex,
    },
    userOptions,
  )
}

function routePlugin(userOptions: UserOptions = {}): Plugin {
  let config: ResolvedConfig | undefined
  let filesPath: string[] = []
  let generatedRoutes: Route[] | null | undefined
  let pagesDirPaths: string[] = []

  const options: ResolvedOptions = resolveOptions(userOptions)

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    configResolved(_config) {
      config = _config
      options.root = config.root
    },
    resolveId(id) {
      if (id === ID)
        return ID
    },
    async load(id) {
      if (id === ID) {
        debug('Loading...')
        debug('Loading Gen Routes: %O', generatedRoutes)

        if (!generatedRoutes) {
          generatedRoutes = []
          filesPath = []
          pagesDirPaths = []
          for (const pageDir of options.pagesDirOptions) {
            const pageDirPath = normalizePath(resolve(options.root, pageDir.dir))
            pagesDirPaths = pagesDirPaths.concat(pageDirPath)
            debug('Loading PageDir: %O', pageDirPath)

            const files = await getFilesFromPath(pageDirPath, options)
            filesPath = filesPath.concat(files.map(f => `${pageDirPath}/${f}`))
            debug('FilesPath: %O', files)

            const routes = generateRoutes(files, pageDir, pageDirPath, options)
            generatedRoutes = generatedRoutes.concat(routes)
            // debug('Routes: %O', generatedRoutes)
          }
        }

        const clientCode = generateClientCode(generatedRoutes, options)

        // debug('Client code: %O', clientCode)

        return clientCode
      }
    },
    async transform(code: string, id: string) {
      const { query } = parseVueRequest(id)

      if (query && query.vue && query.type === 'route') {
        return {
          code: 'export default {}',
          map: null,
        }
      }

      return {
        code,
        map: null,
      }
    },
    async handleHotUpdate({ file, server, read }) {
      const extensionsRE = new RegExp(`\\.(${options.extensions.join('|')})$`)
      const matchedPath = (pagesDirPaths).find(p => file.startsWith(`${p}/`))
      // Handle pages HMR
      if (matchedPath && extensionsRE.test(file)) {
        let needReload = false

        // HMR on new file created
        // Otherwise, handle HMR from custom block
        if (!filesPath.includes(file)) {
          generatedRoutes = null
          needReload = true
        }
        else if (generatedRoutes) {
          const content = await read()
          needReload = updateRouteFromHMR(content, file, generatedRoutes, options)
        }

        if (needReload) {
          const { moduleGraph } = server
          const module = moduleGraph.getModuleById(ID)

          debug('Reload for file: %s', file.replace(options.root, ''))

          return [module] as ModuleNode[]
        }
      }
    },
  }
}

export * from './types'
export default routePlugin
