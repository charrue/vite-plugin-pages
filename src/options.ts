import { UserOptions, ResolvedOptions, PageDirOptions } from './types'

export function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  const {
    pagesDir = ['src/pages'],
    extensions = ['vue', 'js'],
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
    replaceSquareBrackets = false,
    nuxtStyle = false,
    react = false,
  } = userOptions

  const importMode = react ? 'sync' : 'async'

  const root = process.cwd()

  let pagesDirOptions: PageDirOptions[] = []

  if (typeof pagesDir === 'string') {
    pagesDirOptions = pagesDirOptions.concat({ dir: pagesDir, baseRoute: '' })
  }
  else {
    for (const dir of pagesDir) {
      if (typeof dir === 'string')
        pagesDirOptions = pagesDirOptions.concat({ dir, baseRoute: '' })
      else if (dir as PageDirOptions)
        pagesDirOptions = pagesDirOptions.concat(dir)
    }
  }

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`)

  return Object.assign(
    {},
    {
      routeBlockLang,
      root,
      pagesDir,
      pagesDirOptions,
      extensions,
      importMode,
      exclude,
      syncIndex,
      replaceSquareBrackets,
      nuxtStyle,
      react,
      extensionsRE,
    },
    userOptions,
  )
}
