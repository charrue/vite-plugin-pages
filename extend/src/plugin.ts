import { MACROS_FN_STR, MACROS_KEY, isJsx } from './resolveRouteBlock'
import type { Plugin } from 'vite'

export function definePageRoutePlugin(): Plugin {
  return {
    name: 'charrue:definePageRoute',
    enforce: 'pre',

    async transform(code, id) {
      const [filepath] = id.split('?')

      if (isJsx(filepath)) {
        if (!code.includes(MACROS_KEY)) return code
        return `${MACROS_FN_STR}\n${code}`
      }

      return code
    },
  }
}
