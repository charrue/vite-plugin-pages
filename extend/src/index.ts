import { definePageRoutePlugin } from './plugin'
import { resolveRouteBlock } from './resolveRouteBlock'

export { definePageRoutePlugin, resolveRouteBlock }
