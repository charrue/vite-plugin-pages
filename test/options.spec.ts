import { resolveOptions } from '../src/options'

describe('Options', () => {
  test('resolve', () => {
    const options = resolveOptions({
      pagesDir: 'test/assets/pages',
    })
    const reactOptions = resolveOptions({
      pagesDir: 'test/assets/pages',
      react: true,
    })
    delete options.root
    delete reactOptions.root
    expect(options).toMatchSnapshot('resolved options')
    expect(reactOptions).toMatchSnapshot('resolved options - react')
  })
})
