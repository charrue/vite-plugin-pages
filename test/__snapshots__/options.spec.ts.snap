// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options resolve: resolved options - react 1`] = `
Object {
  "exclude": Array [],
  "extendRoute": undefined,
  "extensions": Array [
    "tsx",
    "jsx",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\)\\$/,
  "importMode": "sync",
  "nuxtStyle": false,
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "pages": Array [
    Object {
      "baseRoute": "",
      "dir": "examples/react/src/pages",
    },
  ],
  "replaceSquareBrackets": false,
  "resolver": "react",
  "root": Any<String>,
  "routeBlockLang": "json5",
  "syncIndex": true,
}
`;

exports[`Options resolve: resolved options 1`] = `
Object {
  "exclude": Array [],
  "extendRoute": undefined,
  "extensions": Array [
    "vue",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(vue\\|ts\\|js\\)\\$/,
  "importMode": "async",
  "nuxtStyle": false,
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "pages": Array [
    Object {
      "baseRoute": "",
      "dir": "examples/vue/src/pages",
    },
  ],
  "replaceSquareBrackets": false,
  "resolver": "vue",
  "root": Any<String>,
  "routeBlockLang": "json5",
  "syncIndex": true,
}
`;
